crates/matrix-sdk/src/media.rs

```diff
@@ -140,22 +144,61 @@ impl Media {
     pub async fn get_media_file(
         &self,
         request: &MediaRequest,
+        body: Option<String>,
         content_type: &Mime,
         use_cache: bool,
     ) -> Result<MediaFileHandle> {
         let data = self.get_media_content(request, use_cache).await?;
 
-        let mut suffix = String::from("");
-        if let Some(extension) =
-            mime_guess::get_mime_extensions(content_type).and_then(|a| a.first())
-        {
-            suffix = String::from(".") + extension;
-        }
+        let inferred_extension = mime2ext::mime2ext(content_type);
+
+        let body_path = body.as_ref().map(Path::new);
+        let filename = body_path.and_then(|f| f.file_name().and_then(|f| f.to_str()));
+        let filename_with_extension = body_path.and_then(|f| {
+            if f.extension().is_some() {
+                f.file_name().and_then(|f| f.to_str())
+            } else {
+                None
+            }
+        });
+
+        let (temp_file, temp_dir) = match (filename, filename_with_extension, inferred_extension) {
```

Comment:
I'm not super happy with how the logic is expressed here, but it will probably be easier for me to do a follow-up PR instead of finding the best way to express this now (I'm not sure yet what that would look like).

Reply:
Fair, maybe let me know when you do 👍.  I've made the other changes.